Fix evaluation of decreases clause

SaswatPadhi opened the pull request:
`__CPROVER_decreases` clauses on loops were evaluated outside of loop iterations, which was buggy. In this PR, we fix this behavior: decreases clause is evaluated inside the loop block (after the loop guard is satisfied).

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a ~~The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/~~
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a ~~My commit message includes data points confirming performance improvements (if claimed).~~
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

SaswatPadhi committed:
[code contracts] minor refactor

martin-cs reviewed (approved):
Looks like an appropriate fix.  Thanks for separating out the refactor into it's own commit; that does make reviewing easier.

SaswatPadhi committed:
Fix evaluation of decreases clause on loops

`__CPROVER_decreases` clauses on loops were evaluated outside of loop
iterations (before checking the loop guard), which was buggy.
We fix this behavior in this commit. The clause is now evaluated inside
the loop block (after the loop guard is satisfied).

SaswatPadhi commented:
Thanks for taking a look @martin-cs.
I realized that some of the refactoring (renaming) changes had spilled into the next commit, so I fixed that.

@feliperodri could you please take a look too?

feliperodri commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -38,21 +38,62 @@ Date: February 2016
+  // s1 == l1 && s2 == l2 && s3 == l3>.
+  // In fact, the last element is unnecessary, so we do not create it.
+  exprt::operandst equality_conjunctions(lhs.size());
+  equality_conjunctions[0] = binary_relation_exprt(lhs[0], ID_equal, rhs[0]);
+  for(unsigned int i = 1; i < equality_conjunctions.size() - 1; i++)
+  {
+    binary_relation_exprt component_i_equality{lhs[i], ID_equal, rhs[i]};
+    equality_conjunctions[i] =
+      and_exprt(equality_conjunctions[i - 1], component_i_equality);
+  }
```
```suggestion
  exprt::operandst equality_conjunctions;
  for(unsigned int i = 0; i < lhs.size(); i++)
  {
    binary_relation_exprt component_i_equality{lhs[i], ID_equal, rhs[i]};
    equality_conjunctions.push_back(component_i_equality);
  }
  exprt equality_constraint = conjunction(equality_conjunctions);
```

feliperodri replied:
You'd also need to update the next loop in the same style. Perhaps, even combine them together? Just a suggestion, note a hard requirement.

SaswatPadhi replied:
So the `equality_conjunctions` vector stores a list of conjunctions. In your suggestion, it only stores equality predicates and we compute a single conjunction at the end. That wouldn't work.

Basically we want to have `lhs[0] == rhs[0]`, `lhs[0] == rhs[0] && lhs[1] == rhs[1]`, and so on in the `equality_conjunctions` vector.

feliperodri commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -199,7 +226,23 @@ void code_contractst::check_apply_loop_contracts(
+  // value before the loop
+  if(!decreases_clause.is_nil())
+  {
+    for(size_t i = 0; i < old_decreases_vars.size(); i++)
```
updated previous loops with `size_t` as well for consistency.

SaswatPadhi replied:
Fixed as suggested.

feliperodri commented on src/goto-instrument/contracts/contracts.h:
```diff
@@ -123,7 +123,7 @@ class code_contractst
     goto_functionst::goto_functiont &goto_function,
     const local_may_aliast &local_may_alias,
-    const goto_programt::targett loop_head,
+    goto_programt::targett loop_head,
```
I didn't see any modification in `loop_head`, is this change necessary?

SaswatPadhi replied:
Yes, you can look at line 229 and the `insert_before_swap_and_advance` function (specifically, line 90 in that function).

We insert some code (havoc & assumptions) before the loop head, and this change should have been made earlier. We didn't care about it because we were not really using the `loop_head` after that insertion. Now that we do, we need to advance if correctly.

feliperodri replied:
Thanks for the clarification!

feliperodri reviewed (changes requested):
Only minor changes.
